terminal 'I' to 'S' in gapped alignment breaker in SV-calling step

SHuang-Broad opened the pull request:
@cwhelan , this is the code for turning the edge case where an operator 'I' is the beginning or ending operator in the CIGAR of the contig alignment.
Please kindly review. Thanks!

SHuang-Broad commented:
@cwhelan , turns out this also corrects some mistakes made in mis-interpreting the old code. Just updated.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -145,8 +149,14 @@
-                    // infer reference interval
-                    final SimpleInterval referenceInterval = new SimpleInterval(oneRegion.referenceInterval.getContig(), refIntervalStart, refIntervalStart + memoryCigar.getReferenceLength()-1);
+                    // infer reference interval taking into account of strand
+                    final int refBoundary2ndInTheDirectionOfContig = refBoundary1stInTheDirectionOfContig + (oneRegion.forwardStrand ? memoryCigar.getReferenceLength()-1 : -memoryCigar.getReferenceLength()+1) ;
```
Can you make the last part of the ternary be explicitly `-1 * memoryCigar.getReferenceLength() + 1`?

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -168,8 +178,11 @@
                     // update pointers into reference and contig
-                    refIntervalStart     += op.consumesReadBases() ? memoryCigar.getReferenceLength()  : memoryCigar.getReferenceLength() + operatorLen;
-                    contigIntervalStart  += op.consumesReadBases() ? effectiveReadLen + operatorLen    : effectiveReadLen;
+                    if (oneRegion.forwardStrand)
```
Please add braces here (and to the other if statements in this method)

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -114,20 +113,25 @@
+        int contigIntervalStart = 1 + clippedNBasesFromStart;
         int gapCount = 0;
+        // we are walking along the contig following the cigar, which indicates that we might be walking backwards on the reference if oneRegion.forwardStrand==false
+        int refBoundary1stInTheDirectionOfContig = oneRegion.forwardStrand ? oneRegion.referenceInterval.getStart() : oneRegion.referenceInterval.getEnd();
```
If we reversed the cigar here first, would we be able to get rid of all of the if statements in the body of this loop? It seems like it might make the code a lot cleaner. But I guess you'd still need to keep track of it for the on-contig coordinates. Just something to think about.

SHuang-Broad replied:
The two `if` statements regarding the computation for reference regions would be gone, but the computation for the corresponding contig regions would be made complicated by necessary `if` statements.
inherently, we are dealing with two sets of coordinates which would have opposite directions if the alignment between them is the reverse-strand. So walking along either one following the appropriate representation of the CIGAR would require walking backwards on the other. We might be able to think of a way to code this smartly.

cwhelan replied:
If you can't think of a better way now, that's OK. It just feels like there might be a way to make this simpler eventually.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -182,15 +195,53 @@
         return result;
     }
 
+    private static boolean parseCigarAndConvertTrailingInsertionToSoftClip(final Cigar cigarAlongInput5to3Direction,
```
It's not just 'trailing', right, but either the first or the last operator?

SHuang-Broad replied:
Yep, that's what I mean by "trailing". Any better naming suggestion?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -182,15 +195,53 @@
+            emptyTargetCollection.set(emptyTargetCollection.size()-1, new CigarElement(element.getLength(), CigarOperator.S));
+        }
+
+        return cigarHasInsertionAsTrailingOperator;
```
See my comment earlier asking why we can't just return the new cigar list and get rid of this boolean return value.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusCall.java:
```diff
@@ -133,15 +133,20 @@ static String produceVariantId(final NovelAdjacencyReferenceLocations novelAdjac
-        } else {
-            return breakpoints.complication.insertedSequenceForwardStrandRep.length()
-                    + (breakpoints.complication.dupSeqRepeatNumOnCtg - breakpoints.complication.dupSeqRepeatNumOnRef)*breakpoints.complication.dupSeqForwardStrandRep.length();
+        switch (svtype) {
```
Rather than have switch statements on SVTYPES scattered around the code, it might be nice to create abstract methods on the SVType enum and override them when necessary. That way, when we add a new SV type, we don't have to search for all switch statements and make sure they support the new type correctly. Is that a refactoring you'd be comfortable doing as part of this PR? Other places to modify would include the `produceAlleles` and `produceVariantIds` methods.

SHuang-Broad replied:
Done with a new `SVTYPE` class. It is better now. Thanks!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -27,7 +27,7 @@
     private static final long serialVersionUID = 1L;
     private static final Logger log = LogManager.getLogger(AssemblyAlignmentParser.class);
-    private static final boolean DEBUG_STATS = false;
+    private static final boolean DEBUG_STATS = true;
```
Nope, that was for running with debugging. Will switch back.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -75,8 +75,7 @@
 
-    // TODO: 11/26/16 assuming 'I'/'D' cannot be the beginning or last CIGAR operator AND the following operator cannot be a clipping
-    // TODO: 1/18/17 'I' as first/last operator has been seen in actual data, for the purpose of this function, changing it to soft clipping is fine, but needs to be done later (should be very soon)
+    // TODO: 1/31/17 assuming 'I'/'D' cannot be the operator adjacent to clipping operators 'S'/'H', and 'D' cannot be the beginning or last operator
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -114,20 +113,25 @@
 
-        final List<CigarElement> cigarElements = oneRegion.forwardStrandCigar.getCigarElements();
+        final List<CigarElement> cigarElements = new ArrayList<>();
+        final boolean cigarHasInsertionAsTrailingOperator = parseCigarAndConvertTrailingInsertionToSoftClip(oneRegion.cigarAlong5to3DirectionOfContig, cigarElements);
```
Probably a bad name. By "trailing" I mean the first non-clipping operator from either end.

SHuang-Broad replied:
And because of the assumption that's now enforced in the argument checking step on gap operators not neighboring the clipping operators, "trailing" actually means not the first or the last operator.

SHuang-Broad replied:
I've updated the logic to be permissive about 'I' neighboring 'S'/'H' given that we've seen a SAM record doing it, so the "first non-clipping operator from either end" is what I mean now.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -114,20 +113,25 @@
         final List<CigarElement> cigarMemoryList = new ArrayList<>();
-        final int clippedNBases = SVVariantCallerUtils.getNumClippedBases(true, oneRegion.forwardStrandCigar);
+        int clippedNBasesFromStart = SVVariantCallerUtils.getNumClippedBases(true, oneRegion.cigarAlong5to3DirectionOfContig);
+        if (cigarHasInsertionAsTrailingOperator && cigarElements.get(0).getOperator()==CigarOperator.S) {
```
Done with a new overload of `SVVariantCallerUtils.getNumClippedBases()`. Was trying to avoid new CIGAR construction as much as possible.

SHuang-Broad commented:
@cwhelan , I've addressed the comments in this commit. It's not very small due to the refactoring of `SVTYPE` and dealing with the new edge case cigars like `10S10I10M`.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantCallerUtils.java:
```diff
@@ -38,25 +39,106 @@ static int getTotalHardClipping(final Cigar cigar) {
+     * @param cigarElements
+     */
+    @VisibleForTesting
+    static void checkCigarElements(final List<CigarElement> cigarElements) {
```
`validateCigar` might be a better name?

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusCall.java:
```diff
@@ -172,41 +98,55 @@ static int getSvLength(final NovelAdjacencyReferenceLocations breakpoints, GATKS
+    }
+
+    @VisibleForTesting
+    static Map<String, String> setFlags(final NovelAdjacencyReferenceLocations novelAdjacencyReferenceLocations, final SVTYPE svtype) {
```
It seems like this could also be moved to the SVTYPE subclasses -- what do you think?

SHuang-Broad replied:
You are right! Now I can get rid of the reflections that I disliked. Thanks!

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
As discussed in person, let's start the id with "DEL" or "INS" and then have the repeat expansion or contraction information afterwards.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
Also, as noted in person, DEL shouldn't have an expansion check.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
As noted in person, INS shouldn't have a contraction check.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -145,101 +144,152 @@
+                    // task 3: now add trailing cigar element and create the real cigar for the to-be-returned AR
                     cigarMemoryList.add(new CigarElement(oneRegion.assembledContigLength-contigIntervalEnd-hardClippingAtEnd, CigarOperator.S));
-                    if (hardClippingAtEnd != 0)
+                    if (hardClippingAtEnd != 0) { // be faithful to hard clipping (as the accompanying bases have be hard-clipped away)
```
done.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -145,101 +144,152 @@
 
                     break;
                 default:
                     throw new GATKException("Alignment have N or P in it: " + oneRegion.toString()); // TODO: 1/20/17 still not quite sure if this is quite right, it doesn't blow up on NA12878 WGS, but who knows what happens in the future
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -145,101 +144,152 @@
+        final List<CigarElement> convertedList = convertInsToSoftClipFromOneEnd(cigarElements, true);
+        final List<CigarElement> result = convertInsToSoftClipFromOneEnd(convertedList, false);
+        Utils.validateArg(cigarElements.stream().anyMatch(ele -> ele.getOperator().isAlignment()),
+                "No alignment between ref and ctg after 'I' -> 'S' conversion, this suggest new edge case CIGAR record: " + TextCigarCodec.encode(cigarAlongInput5to3Direction));
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -145,101 +144,152 @@
+        final int numHardClippingBasesFromOneEnd = SVVariantCallerUtils.getNumHardClippingBases(fromStart, cigarElements);
+        final int numSoftClippingBasesFromOneEnd = SVVariantCallerUtils.getNumSoftClippingBases(fromStart, cigarElements);
+
+        final int indexForFirstMaybeInsertion;
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantCallerUtils.java:
```diff
@@ -38,25 +39,106 @@ static int getTotalHardClipping(final Cigar cigar) {
+    @VisibleForTesting
+    static void checkCigarElements(final List<CigarElement> cigarElements) {
+        Utils.validateArg(!cigarElements.isEmpty(), "Cannot parse empty list cigarElements");
+        Utils.validateArg(cigarElements.size()>1 || (cigarElements.size()==1 && cigarElements.get(0).getOperator().isAlignment()),
```
Yes, you are right in that the 4th check covers the 2nd and the 2nd is not a good check. 
The case of `10H10S1D10M` is now caught as well, with new test coverage. Thanks!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantCallerUtils.java:
```diff
@@ -38,25 +39,106 @@ static int getTotalHardClipping(final Cigar cigar) {
+    static void checkCigarElements(final List<CigarElement> cigarElements) {
+        Utils.validateArg(!cigarElements.isEmpty(), "Cannot parse empty list cigarElements");
+        Utils.validateArg(cigarElements.size()>1 || (cigarElements.size()==1 && cigarElements.get(0).getOperator().isAlignment()),
+                "invalid input cigar elements " + cigarElements.toString());
```
case covered by another check now

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantCallerUtils.java:
```diff
@@ -38,25 +39,106 @@ static int getTotalHardClipping(final Cigar cigar) {
+        Utils.validateArg(!cigarElements.isEmpty(), "Cannot parse empty list cigarElements");
+        Utils.validateArg(cigarElements.size()>1 || (cigarElements.size()==1 && cigarElements.get(0).getOperator().isAlignment()),
+                "invalid input cigar elements " + cigarElements.toString());
+        Utils.validateArg ( !(cigarElements.get(0).getOperator().isClipping() && cigarElements.get(1).getOperator()==CigarOperator.D)
```
case covered by new check now

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
Changed to TYPES

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVTYPE.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -145,101 +144,152 @@
-        }
 
-        return cigarHasInsertionAsTrailingOperator;
+            cigarElements.set(indexForFirstMaybeInsertion, new CigarElement(element.getLength(), CigarOperator.S));
```
done

SHuang-Broad commented:
@cwhelan , addressed all comments for the 2nd round. Please take another look.
Thanks!

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -251,47 +247,63 @@
+
+            return compactifyNeighboringSoftClippings(cigarElements);
+
+//            if (numSoftClippingBasesFromOneEnd!=0) { // compactify with possible neighboring clipping
```
Please remove commented out code before checking in.

SHuang-Broad replied:
Sorry for not catching this...

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
```diff
@@ -251,47 +247,63 @@
+        for (final CigarElement element : cigarElements) {
+            if (result.isEmpty() || result.get(result.size()-1).getOperator()!=element.getOperator()) {
+                result.add(element);
+            } else { // no additional checking because this is possible only for two neighboring soft clippings, one of which is converted from an insertion operation
```
Are you sure about that (that this will never happen except for in the case of a converted soft clip)? We could be a little more conservative here and only do the merge if the adjacent elements are in fact soft clips. I just worry that there's some BWA edge case that could give two adjacent "M" operators or something, and this would squish them incorrectly into a new "S" element. Also the name of the method is specific to soft clips.

SHuang-Broad replied:
Done with new exception-throwing check.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
```diff
@@ -16,7 +16,7 @@
     // VCF standard SV header lines
     // todo: add these and the other standard SV info fields from the VCF spec to htsjdk VCFStandardHeaderLines
-    public static final String SVTYPE = "SVTYPE";
+    public static final String SVTYPE = "SvType";
```
I think this still needs to be all uppercase.

cwhelan replied:
Maybe this was an IDE search/replace overstep.

SHuang-Broad replied:
right, IDE overkill...

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVConstants.java:
```diff
@@ -19,10 +19,10 @@ private SVConstants() {
-        static final String VCF_ALT_ALLELE_STRING_INS = "<INS>";
-        static final String VCF_ALT_ALLELE_STRING_DEL = "<DEL>";
-        static final String VCF_ALT_ALLELE_STRING_DUP = "<DUP>";
+        static final String VCF_ALT_ALLELE_STRING_INV = "<Inversion>";
```
These need to be INS, DEL, etc. Maybe more search/replace issues?

SHuang-Broad replied:
yep

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantCallerUtils.java:
```diff
@@ -130,15 +130,37 @@ static int getNumSoftClippingBases(final boolean fromStart, final List<CigarElem
      */
     @VisibleForTesting
-    static void checkCigarElements(final List<CigarElement> cigarElements) {
+    static void validateCigar(final List<CigarElement> cigarElements) {
```
You could update the comment for this method, which still describes the redundant check you removed.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SvType.java:
Should this be "TANDEMUPLICATION_EXPANSION_ID_START_STRING"?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SvType.java:
not sure why i didn't catch this. it is now changed so that tandem duplication is handled completely in the `DuplicationTandem` class.

SHuang-Broad commented:
Holding off because after rebase the aligner is complaining for a null pointer exception. Need to investigate why.

SHuang-Broad commented:
Whole pipeline working now with updated way of running the alignment step, which is reflected in the PR #2435
